### DISCUSSION TITLE
Please share more details about InputNonClientPointerSource

### DISCUSSION THREAD

**Discussion by zhuxb711** [microsoft/WindowsAppSDK#3849] [Q&A] [answered]
`InputNonClientPointerSource` is a new API but without detail documents. According to the document, [AppWindowTitleBar.SetDragRectangles](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindowtitlebar.setdragrectangles?view=windows-app-sdk-1.4) is deprecated. Could share more information and example about how to use it to replace the existing `AppWindowTitleBar.SetDragRectangles`?

**Comment by castorix** [comment]
It is the same code as in MS sample https://learn.microsoft.com/en-us/windows/apps/develop/title-bar?tabs=wasdk#interactive-content
with something like :
```
 IntPtr hWndMain = WinRT.Interop.WindowNative.GetWindowHandle(this);
 Microsoft.UI.WindowId myWndId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWndMain);
 var incps = InputNonClientPointerSource.GetForWindowId(myWndId);
 // in SetDragRegionForCustomTitleBar, MSDN code with dragRectsList... dragRects
 incps.SetRegionRects(NonClientRegionKind.Caption, dragRects);
```

**Reply by zhuxb711** [reply]
I'm not quite understanding the enum "NonClientRegionKind". How to specific those regions we want to include/exclude as draggable regions?

**Reply by DarranRowe** [reply]
@zhuxb711

The simplest way to see it is that if you include it, then it will be part of the non client region. If you don't include anything/remove everything then the clicks will fall through to the underlying window.
This runtime class is a superset of the behaviour in SetDragRects, where SetDragRects would imply that everything that you set is part of the title bar, InputNonClientPointerSource defines what the region you are defining should be seen as. This means that if you want to tune the title bar itself then what you want to do is to call ClearRegionRects with the Caption and Icon kinds, and then set your regions, or set the region you want and give it the Passthrough kind.
It becomes easier if you remember that this is no longer just the title bar, but instead the entire frame, border and title bar region. It also helps if you know how to use DWM to draw your own non client region, since this should be equivalent to the custom hit testing that you have to do.

@castorix

To make things a bit easier, remember that since 1.3, Microsoft.UI.Xaml.Window has an AppWindow property. Microsoft.UI.Windowing.AppWindow has an Id property.

**Comment by zhuxb711** [comment, marked as answer]
@DarranRowe Thanks.
Here is the code that helps set the NonClient region using `InputNonClientPointerSource`. Please correct me if any misunderstanding. 
Hope the code could help someone who seeking the solution on `InputNonClientPointerSource`

```csharp
public static void SetNonClientRegionElementsOnTitleBar(this Window Window, FrameworkElement Element, params FrameworkElement[] ExceptElements)
{
    if (ExceptElements.Contains(Element))
    {
        throw new ArgumentException($"\"{nameof(ExceptElements)}\" contains the \"{nameof(Element)}\" which is not allowed", nameof(Element));
    }

    float DpiScaleFactor = User32.GetDpiScaleFactorFromWindow(Window.GetWindowHandle());

    if (DpiScaleFactor > 0)
    {
        Point ElementPoint = Element.TransformToVisual(Window.Content).TransformPoint(new Point(0, 0));
        RectInt32 DraggableRect = new RectInt32((int)(ElementPoint.X * DpiScaleFactor), (int)(ElementPoint.Y * DpiScaleFactor), (int)(Element.ActualWidth * DpiScaleFactor), (int)(Element.ActualHeight * DpiScaleFactor));

        if (InputNonClientPointerSource.GetForWindowId(Window.AppWindow.Id) is InputNonClientPointerSource NonClientSource)
        {
            NonClientSource.ClearRegionRects(NonClientRegionKind.Caption);
            NonClientSource.SetRegionRects(NonClientRegionKind.Caption, new RectInt32[] { DraggableRect });

            if (ExceptElements.Length > 0)
            {
                List<RectInt32> ExceptRects = new List<RectInt32>(ExceptElements.Length);

                foreach (FrameworkElement ExceptElement in ExceptElements)
                {
                    Point ExceptElementPoint = ExceptElement.TransformToVisual(Window.Content).TransformPoint(new Point(0, 0));
                    ExceptRects.Add(new RectInt32((int)(ExceptElementPoint.X * DpiScaleFactor), (int)(ExceptElementPoint.Y * DpiScaleFactor), (int)(ExceptElement.ActualWidth * DpiScaleFactor), (int)(ExceptElement.ActualHeight * DpiScaleFactor)));
                }

                NonClientSource.ClearRegionRects(NonClientRegionKind.Passthrough);
                NonClientSource.SetRegionRects(NonClientRegionKind.Passthrough, ExceptRects.Distinct().ToArray());
            }
        }
    }
}
```
